packages/utils/src/ClipboardUtils.ts

```diff
@@ -43,39 +48,51 @@ export function copyToClipboardExecCommand(text: string): void {
 
 /**
  * Reads text from the clipboard.
- * @returns Promise that resolves to text from clipboard as string, or null if permissions are not supported/granted.
+ * @returns Promise that resolves to text from clipboard as string, or rejects if permissions are not supported/granted.
  */
-export async function readFromClipboard(): Promise<string | null> {
-  const { clipboard } = navigator;
-  if (clipboard === undefined) return null;
-
-  let permissionState = await checkPermission('clipboard-read');
-
-  if (permissionState === 'granted') {
-    const text = await clipboard.readText();
-    return text;
-  }
+export async function readFromClipboard(): Promise<string> {
+  try {
+    const { clipboard } = navigator;
+    if (clipboard === undefined) throw new ClipboardUnavailableError();
 
-  if (permissionState === 'prompt') {
-    try {
-      // Need to call this to bring up a permission prompt
-      await clipboard.readText();
-    } catch {
-      // Ignore error caused by calling readText() without permissions
-    }
+    let permissionState = await checkPermission('clipboard-read');
 
-    // Check status again after user has interacted with the permission prompt
-    permissionState = await checkPermission('clipboard-read');
     if (permissionState === 'granted') {
       const text = await clipboard.readText();
       return text;
     }
 
-    if (permissionState === 'prompt' || permissionState === 'denied') {
-      // Prompt means user closed out of the previous permission prompt, treat it as a denial
-      return null;
+    if (permissionState === 'prompt') {
+      try {
+        // Need to call this to bring up a permission prompt
+        await clipboard.readText();
+      } catch {
+        // Ignore error caused by calling readText() without permissions
+      }
+
+      // Check status again after user has interacted with the permission prompt
+      permissionState = await checkPermission('clipboard-read');
+      if (permissionState === 'granted') {
+        const text = await clipboard.readText();
+        return text;
+      }
+
+      if (permissionState === 'prompt' || permissionState === 'denied') {
+        // Prompt means user closed out of the previous permission prompt, also treat it as a denial
+        throw new ClipboardPermissionsDeniedError();
+      }
+    }
+
+    if (permissionState === 'denied') {
+      throw new ClipboardPermissionsDeniedError();
+    }
```

Comment:
You can simplify this this:
```suggestion
    }

    if (permissionState === 'prompt' || permissionState === 'denied') {
        // Prompt means user closed out of the previous permission prompt, also treat it as a denial
      throw new ClipboardPermissionsDeniedError();
    }
```